Jenkinsfile

```diff
@@ -1,89 +1,162 @@
+def prepareMachineDirectory(FileManager fm, DirKeeper dk, FileNameKeeper fnk) {
+    fm.newDir(dk.machineDir())
+    sh "chmod 500 ${dk.provisionDir()}/${fnk.prepareTemplateName()}"
+    fm.copy(fnk.prepareTemplateName(), dk.provisionDir(), dk.machineDir())
+    copyTerraformFiles(fm, dk, fnk)
+    copyDaemonFile(fm, dk, fnk)
+    resolveInstancesYamlFile(fm, dk, fnk)
+    prepareKeyFiles(fm, dk, fnk)
 }
 
-def copyTerraformFiles() {
-    sh "cp provision/${params.orchestrator}/${params.orchestrator}.tf ../${mainDirectoryName}/${params.Login}/${params.MachineName}/${params.orchestrator}.tf"
-    sh "cp provision/${params.orchestrator}/variables.tf ../${mainDirectoryName}/${params.Login}/${params.MachineName}/variables.tf"
+def prepareKeyFiles(FileManager fm, DirKeeper dk, FileNameKeeper fnk) {
+    if (fnk.userPubKeyName() != "" && fnk.userPrivKeyName() != "") {
+        fm.move(fnk.userPubKeyName(), "${dk.machineDir()}/${fnk.mainPubKeyName()}")
+        fm.move(fnk.userPrivKeyName(), "${dk.machineDir()}/${fnk.mainPrivKeyName()}")
+    } else {
+        warningEcho("Didn't get Public and Private SSH key. Using default keys.")
+        fm.copy("${dk.provisionDir()}/${fnk.defaultPubKeyName()}", "${dk.machineDir()}/${fnk.mainPubKeyName()}")
+        fm.copy("${dk.provisionDir()}/${fnk.defaultPrivKeyName()}", "${dk.machineDir()}/${fnk.mainPrivKeyName()}")
+    }
+    sh "chmod 600 ${dk.machineDir()}/${fnk.mainPubKeyName()}"
+    sh "chmod 600 ${dk.machineDir()}/${fnk.mainPrivKeyName()}"
 }
 
-def copyDaemonFile() {
-    sh "cp provision/daemon.json ../${mainDirectoryName}/${params.Login}/${params.MachineName}/daemon.json"
+def setDirKeeper(DirKeeper dk) {
+    dk.setMainDir("../${Constants.mainDirectoryName}")
+    dk.setUserDir("../${Constants.mainDirectoryName}/${params.Login}")
+    dk.setMachineDir("../${Constants.mainDirectoryName}/${params.Login}/${params.MachineName}")
+    dk.setProvisionDir("provision")
+    dk.setTfFilesDir("provision/${params.orchestrator}")
 }
 
-def resolveInstancesYamlFile() {
-    if ("${instances_yaml}" != "") {
-        def instancesYaml = unstashParam "instances_yaml"
-        sh "mv ${instancesYaml} ../${mainDirectoryName}/${params.Login}/${params.MachineName}/template.yaml"
-    } else {
-        warningEcho("Didn't get instances yaml file. Using default file.")
-        sh "cp provision/template.yaml ../${mainDirectoryName}/${params.Login}/${params.MachineName}/template.yaml"
+def setFileNameKeeper(FileNameKeeper fnk, String operation) {
+    fnk.setDaemonName("daemon.json")
+    fnk.setPrepareTemplateName("prepare_template")
+    fnk.setTfVarsName("variables.tf")
+    fnk.setTfMainName("${params.orchestrator}.tf")
+
+    fnk.setDefaultPubKeyName("id_rsa.pub")
+    fnk.setDefaultPrivKeyName("id_rsa")
+    fnk.setDefaultInstanceName("template.yaml")
+
+    fnk.setMainPubKeyName("${params.Login}-key.pub")
+    fnk.setMainPrivKeyName("${params.Login}-key.priv")
+    fnk.setMainInstanceName("instances.yaml")
```

Comment:
Consider the FileNameKeeper class. It is a bunch of setters, but in reality it takes only two parameters: params.orchestrator and params.login.

What is needed here, is a constructor with two parameters, not millions of setters/getters. They break encapsulation anyways, so right now we have a struct with extra steps...

If a struct for keeping all the filenames is required, then make it a struct. This goes for other classes - they have no functionality and should probably be bare structs, with eveything hardcoded in their constructors except of one or two params.